src/engine/ebur128analysis.cpp

```diff
@@ -0,0 +1,397 @@
+/*
+ * Strawberry Music Player
+ * Copyright 2023 Roman Lebedev
+ *
+ * Strawberry is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Strawberry is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Strawberry.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "config.h"
+
+#include <cmath>
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <optional>
+
+#include <ebur128.h>
+#include <glib-object.h>
+#include <glib.h>
+#include <gst/app/gstappsink.h>
+#include <gst/gst.h>
+
+#include <QCoreApplication>
+#include <QElapsedTimer>
+#include <QString>
+#include <QThread>
+#include <QtGlobal>
+
+#include "core/logging.h"
+#include "core/signalchecker.h"
+
+#include "ebur128analysis.h"
+
+static const int kTimeoutSecs = 60;
+
+namespace {
+
+struct ebur128_state_deleter {
+  void operator()(ebur128_state *p) const { ebur128_destroy(&p); };
+};
+
+struct GstSampleDeleter {
+  void operator()(GstSample *s) const { gst_sample_unref(s); };
+};
+
+struct FrameFormat {
+  enum class DataFormat {
+    S16,
+    S32,
+    FP32,
+    FP64
+  };
+
+  int channels;
+  int samplerate;
+  DataFormat format;
+
+  explicit FrameFormat(GstCaps *caps);
+};
+
+class EBUR128State {
+ public:
+  EBUR128State() = delete;
+  EBUR128State(const EBUR128State&) = delete;
+  EBUR128State(EBUR128State&&) = delete;
+
+  EBUR128State &operator=(const EBUR128State&) = delete;
+  EBUR128State &operator=(EBUR128State&&) = delete;
+
+  explicit EBUR128State(FrameFormat dsc_);
+  const FrameFormat dsc;
+
+  void AddFrames(const char *data, size_t size);
+
+  static std::optional<EBUR128Measures> Finalize(EBUR128State &&state);
+
+ private:
+  std::unique_ptr<ebur128_state, ebur128_state_deleter> st;
+};
+
+class EBUR128AnalysisImpl {
+  EBUR128AnalysisImpl() = default;
+
+ public:
+  static std::optional<EBUR128Measures> Compute(const Song &song);
+
+ private:
+  GstElement *convert_element_;
+
+  std::optional<EBUR128State> state;
+
+  static void NewPadCallback(GstElement *elt, GstPad *pad, gpointer data);
+  static GstFlowReturn NewBufferCallback(GstAppSink *app_sink, gpointer self);
+};
+
+FrameFormat::FrameFormat(GstCaps *caps) {
+
+  GstStructure *structure = gst_caps_get_structure(caps, 0);
+  QString format_str = gst_structure_get_string(structure, "format");
+  gst_structure_get_int(structure, "rate", &samplerate);
+  gst_structure_get_int(structure, "channels", &channels);
```

Comment:
One thing i didn't do, is channel mapping.
It is a bit confusing because libebur128 operates in terms of `ITU R-REC-BS 1770-4`,
while GStreamer uses `SMPTE 2036-2-2008`.
In vast majority of cases it will not matter, but it will matter for exotic cases.